db/db_test.go

```diff
@@ -226,3 +233,58 @@ func TestSeek(t *testing.T) {
 		return nil
 	})
 }
+
+func TestConcurrentUpdate(t *testing.T) {
+	testDb := db.NewTestDb()
+	defer testDb.Close()
+	wg := sync.WaitGroup{}
+
+	key := []byte{0}
+	require.NoError(t, testDb.Update(func(txn db.Transaction) error {
+		return txn.Set(key, []byte{0})
+	}))
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for i := 0; i < 10; i++ {
+				assert.NoError(t, testDb.Update(func(txn db.Transaction) error {
+					var next byte
+					err := txn.Get(key, func(bytes []byte) error {
+						next = bytes[0] + 1
+						return nil
+					})
+					if err != nil {
+						return err
+					}
+					return txn.Set(key, []byte{next})
+				}))
+			}
+		}()
+	}
+
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for i := 0; i < 10; i++ {
+				txn := testDb.NewTransaction(true)
+				var next byte
+				require.NoError(t, txn.Get(key, func(bytes []byte) error {
+					next = bytes[0] + 1
+					return nil
+				}))
+				require.NoError(t, txn.Set(key, []byte{next}))
+				txn.Commit()
+			}
+		}()
+	}
+
+	wg.Wait()
+	require.NoError(t, testDb.View(func(txn db.Transaction) error {
+		return txn.Get(key, func(bytes []byte) error {
+			assert.Equal(t, byte(200), bytes[0])
+			return nil
+		})
+	}))
+}
```

Comment:
Please split this case into 2 separate sub tests. One should be called `t.Run("Update()", ...)` and the other `t.Run("NewTransaction()", ...) `

Reply:
It is intentional that they are in the same test tho. We dont really mix Update and NewTransaction in our code. But they should be respecting only-one-update-at-a-time rule.

ie NewTransaction(true) should block if a Update() is running and vice versa.